src/NHibernate/Event/EntityListener.cs

```diff
@@ -0,0 +1,211 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.Serialization;
+using NHibernate.Persister.Entity;
+using NHibernate.Util;
+
+namespace NHibernate.Event
+{
+    internal class EventConfig
+    {
+        public object Entity { get; set; }
+
+        public IEntityPersister Persister { get; set; }
+
+        public IEventSource Session { get; set; }
+
+        public object[] State { get; set; }
+
+        public bool ShouldUpdateState
+        {
+            get { return State != null; }
+        }
+    }
+
+    [Serializable]
+    public class EntityListener : IPreUpdateEventListener, IPreInsertEventListener, IPreDeleteEventListener,
+        IFlushEntityEventListener, IPostUpdateEventListener, IPostInsertEventListener, IPostDeleteEventListener
+    {
+        [NonSerialized]
+        private Dictionary<System.Type, Dictionary<System.Type, IEnumerable<MethodInfo>>>
+            _methodsWithListenerAttributeCache;

```

Comment:
Instances of this class are potentially shared among many threads. This dictionary must be changed for a `ConcurrentDictionary`.